assayist/client/query.py

```diff
@@ -44,3 +46,79 @@ def get_container_content_sources(build_id):
         'internal_urls': internal_urls,
         'upstream_urls': upstream_urls
     }
+
+
+def get_current_and_previous_versions(name, type_, version):
+    """
+    Find the current and previous source locations.
+
+    :param str name: the canonical name of the component
+    :param str type_: the canonical type of the component
+    :param str version: the canonical version of the source location
+    :return: a dictionary of all the previous source locations and the current source location
+    :rtype: dict
+    """
+    # TODO: Consider alternative names as well
+    query = """
+        MATCH (:Component {{canonical_name: '{name}', canonical_type: '{type}'}})
+            <-[:SOURCE_FOR]-(:SourceLocation {{canonical_version: '{version}'}})
+            -[:SUPERSEDES*0..]->(sl:SourceLocation)
+        RETURN sl
+    """.format(name=name, type=type_, version=version)
+    results, _ = neomodel.db.cypher_query(query)
+    rv = []
+    for result in results:
+        rv.append(dict(result[0]))
+    return rv
+
+
+def get_container_built_with_sources(source_locations):
+    """
+    Match container builds that used the input source locations to build the content in the image.
+
+    :param list source_locations: a list of source location dictionaries to match against
+    :return: a list of affected container build Koji IDs
+    :rtype: list
+    """
+    if not source_locations or not isinstance(source_locations, collections.Iterable):
+        raise InvalidInput('The input must be a list of source locations')
+
+    query = """
+        // First get all the input source locations
+        MATCH (input_sl:SourceLocation) WHERE input_sl.url IN ['{0}']
+        // Then find all the source locations that embed the input source locations recursively.
+        // The resulting `sl` variable has all the input source locations and all the source
+        // locations that eventually embed the input source locations.
```

Comment:
Actually, we don't want directionless because that means it'll include the source locations that the input source locations embed. We only want the other way around. I'll make the upstream and embeds queries separate.